src/lang/modifyAst/faces.test.ts

```diff
@@ -0,0 +1,357 @@
+import {
+  type Artifact,
+  type ArtifactGraph,
+  assertParse,
+  recast,
+} from '@src/lang/wasm'
+import type { Selection, Selections } from '@src/lib/selections'
+import { enginelessExecutor } from '@src/lib/testHelpers'
+import { err } from '@src/lib/trap'
+import {
+  addShell,
+  retrieveFaceSelectionsFromOpArgs,
+} from '@src/lang/modifyAst/faces'
+import { initPromise } from '@src/lang/wasmUtils'
+import { engineCommandManager, kclManager } from '@src/lib/singletons'
+import { getCodeRefsByArtifactId } from '@src/lang/std/artifactGraph'
+import { createPathToNodeForLastVariable } from '@src/lang/modifyAst'
+import { stringToKclExpression } from '@src/lib/kclHelpers'
+import type { KclCommandValue } from '@src/lib/commandTypes'
+import env from '@src/env'
+
+// Unfortunately, we need the real engine here it seems to get sweep faces populated
+beforeAll(async () => {
+  await initPromise
+  await new Promise((resolve) => {
+    engineCommandManager.start({
+      token: env().VITE_KITTYCAD_API_TOKEN,
+      width: 256,
+      height: 256,
+      setMediaStream: () => {},
+      setIsStreamReady: () => {},
+      callbackOnEngineLiteConnect: () => {
+        resolve(true)
+      },
+    })
+  })
+}, 30_000)
+
+afterAll(() => {
+  engineCommandManager.tearDown()
+})
+
+async function getAstAndArtifactGraph(code: string) {
+  const ast = assertParse(code)
+  await kclManager.executeAst({ ast })
+  const {
+    artifactGraph,
+    execState: { operations },
+  } = kclManager
+  await new Promise((resolve) => setTimeout(resolve, 100))
+  return { ast, artifactGraph, operations }
+}
+
+function createSelectionFromArtifacts(
+  artifacts: Artifact[],
+  artifactGraph: ArtifactGraph
+): Selections {
+  const graphSelections = artifacts.flatMap((artifact) => {
+    const codeRefs = getCodeRefsByArtifactId(artifact.id, artifactGraph)
+    if (!codeRefs || codeRefs.length === 0) {
+      return []
+    }
+
+    return {
+      codeRef: codeRefs[0],
+      artifact,
+    } as Selection
+  })
+  return {
+    graphSelections,
+    otherSelections: [],
+  }
+}
+
+async function getAstAndSolidSelections(code: string) {
+  const { ast, artifactGraph } = await getAstAndArtifactGraph(code)
+  const artifacts = [...artifactGraph.values()].filter((a) => a.type === 'path')
+  if (artifacts.length === 0) {
+    throw new Error('Artifact not found in the graph')
+  }
+  const solids = createSelectionFromArtifacts(artifacts, artifactGraph)
+  return { artifactGraph, ast, solids }
+}
```

Comment:
These seem potentially generic enough to become test helpers in the future, just want to flag it but I support leaving it here for now.

Reply:
Yup! I expect this faces file to also contain hole stuff later, which is why I named it generically like that, that's the door I'm trying to keep open. Or maybe other things like drafts.